### PR TITLE
Implement V2 Map Generation

### LINKED ISSUE #1: Implement Map Generation
The map generation system generates the map, populating the initial entities.

### LINKED ISSUE #59: Unable to directly return to previous room
Introduced in #56 is a bug where after entering a new room, you have to take a step into the doorway before you use it again. You cannot directly go back and forth between rooms with a single key press.

### PR DESCRIPTION
**Pull request by 6167656e74323431**
Implement the new proper map generation.

Fixes #1.
Fixes #59.

Current progress:

- [x] Room layout generation
- [x] Random chest population
- [x] Random battle generation
- [x] Checks that make sure it's actually beatable (<- hardest one)

### CODE REVIEW

**Review by sbrad044 — APPROVED** [review]

**Review by zhaoan12 — DISMISSED** [review]
Nice classes

File: src/main/java/com/mg105/use_cases/MapGenerator.java [hunk 1]
```diff
@@ -2,73 +2,366 @@
...
+
+        // By generating two booleans, there is a 50% chance of upgrade token, 25% chance of each potion.
+        if (random.nextBoolean()) {
+            reward = new UpgradeToken();
+        } else if (random.nextBoolean()) {
+            reward = new HealthPotion();
+        } else {
+            reward = new MegaPotion();
+        }
+
```
Comment (siddharthgowda) [inline src/main/java/com/mg105/use_cases/MapGenerator.java:270]: its might be better to use ItemFactory here

**Review by m-grigoriu — APPROVED** [review]

**Review by sbrad044 — DISMISSED** [review]

File: move_data.csv [hunk 2]
```diff
@@ -0,0 +1,9 @@
+name,moveChange,damageChange,isFriendly
+Slow swing,-3,0,FALSE
+Nullify,0,-1,FALSE
+Strong swing,-4,0,FALSE
+Strong heal,6,0,TRUE
+Weak heal,3,0,TRUE
+Reinforce,2,2,TRUE
+Surprise attack,-5,0,FALSE
+Sabotage,-2,-2,FALSE
```
Comment (hoodiexxx) [inline move_data.csv:9]: good csv setup

**Review by hoodiexxx — APPROVED** [review]

**Review by siddharthgowda — APPROVED** [review]
LGTM only thing is that u could use an ItemFactory for adding items to chests (also you change the way healable moves are formed )

**Review by m-grigoriu — APPROVED** [review]

**Review by zhaoan12 — APPROVED** [review]
Nice classes and tests

**Pull request merged**